src/recordhost.cpp

```diff
@@ -0,0 +1,342 @@
+        break;
+    case AttachToProcess:
+        if (m_pids.isEmpty())
+            return false;
+        break;
+    case ProfileSystem:
+        break;
+    case NUM_RECORD_TYPES:
+        Q_ASSERT(false);
+    }
+
+    return m_error.isEmpty()
+        && std::none_of(jobs.begin(), jobs.end(), [](const JobTracker* job) { return job->isJobRunning(); });
+}
+
+void RecordHost::setHost(const QString& host)
+{
+    Q_ASSERT(QThread::currentThread() == thread());
+
+    // don't refresh if on the same host
+    if (host == m_host)
+        return;
+
+    emit isReadyChanged(false);
+
+    m_host = host;
+    emit hostChanged();
+
+    // invalidate everything
+    m_cwd.clear();
+    emit currentWorkingDirectoryChanged(m_cwd);
+
+    m_clientApplication.clear();
+    emit clientApplicationChanged(m_clientApplication);
+
+    m_perfCapabilities = {};
+    emit perfCapabilitiesChanged(m_perfCapabilities);
+
+    m_checkPerfCapabilitiesJob.startJob([](auto&&) { return fetchPerfCapabilities(); },
+                                        [this](RecordHost::PerfCapabilities capabilities) {
+                                            Q_ASSERT(QThread::currentThread() == thread());
+
+                                            m_perfCapabilities = capabilities;
+                                            emit perfCapabilitiesChanged(m_perfCapabilities);
+                                        });
+
+    auto perfPath = Settings::instance()->perfPath();
+    m_checkPerfInstalledJob.startJob(
+        [isLocal = isLocal(), perfPath](auto&&) {
+            if (isLocal) {
+                if (perfPath.isEmpty()) {
+                    return !QStandardPaths::findExecutable(QStringLiteral("perf")).isEmpty();
+                }
+
+                return QFileInfo::exists(perfPath);
+            }
+
+            qWarning() << "remote is not implemented";
+            return false;
+        },
+        [this](bool isInstalled) {
+            if (!isInstalled) {
+                emit errorOccurred(tr("perf is not installed"));
+            }
+            m_isPerfInstalled = isInstalled;
+            emit isPerfInstalledChanged(isInstalled);
+        });
+}
+
+void RecordHost::setCurrentWorkingDirectory(const QString& cwd)
+{
+    Q_ASSERT(QThread::currentThread() == thread());
+
+    if (isLocal()) {
+        QFileInfo folder(cwd);
+
+        if (!folder.exists()) {
+            emit errorOccurred(tr("Working directory folder cannot be found: %1").arg(cwd));
+        } else if (!folder.isDir()) {
+            emit errorOccurred(tr("Working directory folder is not valid: %1").arg(cwd));
+        } else if (!folder.isWritable()) {
+            emit errorOccurred(tr("Working directory folder is not writable: %1").arg(cwd));
+        } else {
+            emit errorOccurred({});
+            m_cwd = cwd;
+            emit currentWorkingDirectoryChanged(cwd);
+        }
+        return;
+    }
+
+    qWarning() << __FUNCTION__ << "is not implemented for remote";
```

Comment:
piping `__FUNCTION__` is useless, just setup `QT_MESSAGE_PATTERN` in your dev env with `%{function}` and you'll get that for free